rfcs/0000-docstrings.md

```diff
@@ -0,0 +1,101 @@
+- Feature Name: docstrings
+- Start Date: 2020-06-27
+- RFC PR: [FuelLabs/sway-rfcs#0000](https://github.com/FuelLabs/sway-rfcs/pull/2)
+- Sway Issue: [FueLabs/sway#149](https://github.com/FuelLabs/Sway/issues/149)
+
+# Summary
+
+[summary]: #summary
+
+Sway will convert Rust-style docstrings (`///`-style comments) into a `#[doc("documentation string")]` annotation. This annotation can then be consumed by tooling and plugins to produce
+rendered documentation, a la `cargo doc` or `docs.rs`.
+
+# Motivation
+
+[motivation]: #motivation
+
+As Sway's designs are largely motivated by Rust's design, we have always wanted to support in-code documentation in this manner. In-code documentation support has been part of the Sway vision
+since the beginning.
+
+This particular approach is both low-friction, in that it utilizes our existing features (annotations, comments, forc plugins); and consistent with Rust's implementation of docstrings. Upon
+the implementation of this RFC in Sway, forc plugins should be able to produce documentation and documentation-related tooling based on in-code docstrings.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A _docstring_ is a line of prose within a Sway source code file. This plain-language string serves to provide exposition and contextual explanation to the subsequent line of Sway code.
+
+Docstrings, for convenience, may be written as a comment with three slashes. E.g.:
+
+```rust
+/// The entry point to the script.
+fn main() {}
+```
+
+The above code example shows a _docstring_ documenting a function item, `fn main()`.
+
+Generally, Sway programmers should find this to be the most expedient and convenient way to provide API documentation for libraries and contract ABIs, and documentation of internal concepts within scripts and predicates.
+
+This style of docstring should be familiar to Rust programmers, and the concept of docstrings is generally prevalent in modern programming languages and should not require significant explanation or new educational material.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Docstrings should be implemented as attributes (a.k.a. annotations) ([reference 1](https://github.com/FuelLabs/sway/issues/470), [reference 2](https://github.com/FuelLabs/sway/pull/1518), [reference 3 (rust)](https://doc.rust-lang.org/reference/attributes.html)). Because annotations/attributes can be applied to any `Item` in Sway ([reference 1](https://github.com/FuelLabs/sway/blob/master/sway-parse/src/attribute.rs#L4), [reference 2](https://github.com/FuelLabs/sway/blob/ba30e8e5ccbb0512aacbaee594473da9e0839c3d/sway-parse/src/item/mod.rs#L13)), this means that any `Item` can be documented with this feature.
```

Comment:
This would also be a great reference to keep track of:

https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html